File: lab1_go/containers/queue.go
package containers

import "fmt"

// QueueNode представляет узел очереди
type QueueNode struct {
	Data string
	Next *QueueNode
}

// Queue представляет очередь
type Queue struct {
	Front *QueueNode
	Rear  *QueueNode
}

// NewQueue создает новую очередь
func NewQueue() *Queue {
	return &Queue{Front: nil, Rear: nil}
}

// Enqueue добавляет элемент в конец очереди
func (q *Queue) Enqueue(value string) {
	newNode := &QueueNode{Data: value, Next: nil}
	
	if q.Rear == nil {
		q.Front = newNode
		q.Rear = newNode
	} else {
		q.Rear.Next = newNode
		q.Rear = newNode
	}
}

// Dequeue удаляет и возвращает элемент из начала очереди
func (q *Queue) Dequeue() (string, error) {
	if q.IsEmpty() {
		return "", fmt.Errorf("Ошибка: очередь пуста")
	}

	value := q.Front.Data
	q.Front = q.Front.Next
	
	if q.Front == nil {
		q.Rear = nil
	}
	
	return value, nil
}

// IsEmpty проверяет, пуста ли очередь
func (q *Queue) IsEmpty() bool {
	return q.Front == nil
}

// Read выводит все элементы очереди
func (q *Queue) Read() {
	current := q.Front
	for current != nil {
		fmt.Print(current.Data, " ")
		current = current.Next
	}
	fmt.Println()
}

// ToSlice преобразует очередь в срез строк
func (q *Queue) ToSlice() []string {
	result := []string{}
	current := q.Front
	for current != nil {
		result = append(result, current.Data)
		current = current.Next
	}
	return result
}

// Clear очищает очередь
func (q *Queue) Clear() {
	q.Front = nil
	q.Rear = nil
}

File: lab1_cpp/scripts/containers/queue.cpp
#include "queue.h"
#include <iostream>
#include <fstream>

using namespace std;

// Конструктор очереди
Queue::Queue() : front(nullptr), rear(nullptr) {}

// Деструктор очереди
Queue::~Queue() {
    clearQueue(*this);
}

// Добавляет элемент в очередь
void enqueue(Queue& queue, const string& value) {
    QueueNode* newNode = new QueueNode{value, nullptr};
    if (queue.rear == nullptr) {
        queue.front = queue.rear = newNode;
    } else {
        queue.rear->next = newNode;
        queue.rear = newNode;
    }
}

// Извлекает элемент из очереди
string dequeue(Queue& queue) {
    if (isQueueEmpty(queue)) {
        throw runtime_error("Ошибка: очередь пуста");
    }

    QueueNode* temp = queue.front;
    string value = temp->data;
    queue.front = temp->next;

    if (queue.front == nullptr) {
        queue.rear = nullptr;
    }

    delete temp;
    return value;
}

// Проверяет, пуста ли очередь
bool isQueueEmpty(const Queue& queue) {
    return queue.front == nullptr;
}

// Очищает очередь
void clearQueue(Queue& queue) {
    while (!isQueueEmpty(queue)) {
        dequeue(queue);
    }
}

// Сохраняет очередь в файл
void saveQueueToFile(const Queue& queue, const string& filePath) {
    ofstream outFile(filePath);
    if (!outFile) {
        throw runtime_error("Ошибка: невозможно открыть файл для записи");
    }
    QueueNode* current = queue.front;
    while (current != nullptr) {
        outFile << current->data << endl;
        current = current->next;
    }
    outFile.close();
}

// Загружает очередь из файла
void loadQueueFromFile(Queue& queue, const string& filePath) {
    ifstream inFile(filePath);
    if (!inFile) {
        throw runtime_error("Ошибка: невозможно открыть файл для чтения");
    }
    clearQueue(queue);
    string value;
    while (getline(inFile, value)) {
        enqueue(queue, value);
    }
    inFile.close();
}